dm-agent/src/main.rs

```diff
@@ -0,0 +1,354 @@
+use futures::future::FutureExt;
+use mqtt_client::{Client, Message, MessageStream, QoS, Topic, TopicFilter};
+use std::sync::Arc;
+use std::{convert::TryFrom, str::from_utf8};
+use thin_edge_json::{
+    group::MeasurementGrouper,
+    measurement::current_timestamp,
+    measurement::FlatMeasurementVisitor,
+    serialize::{ThinEdgeJsonSerializationError, ThinEdgeJsonSerializer},
+};
+use tokio::{select, time::Duration};
+use tracing::{debug_span, error, info, instrument, Instrument};
+
+const DEFAULT_STATS_COLLECTION_WINDOW: u64 = 1000;
+
+const DEFAULT_HOST: &str = "localhost";
+const DEFAULT_PORT: u16 = 1883;
+const CLIENT_ID: &str = "tedge-dm-agent";
+const SOURCE_TOPIC: &str = "collectd/#";
+const TARGET_TOPIC: &str = "tedge/measurements";
+const METRIC_GROUP_KEY_POS: usize = 2;
+const METRIC_KEY_POS: usize = 3;
+
+const APP_NAME: &str = "tedge-dm-agent";
+const DEFAULT_LOG_LEVEL: &str = "warn";
+const TIME_FORMAT: &str = "%Y-%m-%dT%H:%M:%S%.3f%:z";
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    MqttError(#[from] Arc<mqtt_client::Error>), //TODO: avoid the logical duplication of this error and MqttClientError because of the Arc
+
+    #[error(transparent)]
+    MqttClientError(#[from] mqtt_client::Error),
+
+    #[error(
+        "Message received on invalid collectd topic: {0}. \
+        Collectd message topics must be in the format collectd/<hostname>/<metric-plugin-name>/<metric-key>"
+    )]
+    InvalidMeasurementTopic(String),
+
+    #[error("Non UTF-8 payload received on topic: {0}")]
+    InvalidMeasurementPayload(String),
+
```

Comment:
Are all these errors required?

* It looks like as some copy/paste from other parts of the code.
* I would expect a single `#[error(transparent)]  ThinEdgeJsonSerializationError `.
* A refactoring of the thin-edge-json parser error type might be required.

Reply:
These are collectd payload related errors not Thin Edge JSON payload related. For all the Thin Edge JSON errors there's a single error type already defined, called `ThinEdgeJsonSerializationError` exactly as you've suggested.